From e66f92446974ff8d13713836f6e7d2d472aca209 Mon Sep 17 00:00:00 2001
From: Mike Hsieh <mike_hsieh@quantatw.com>
Date: Thu, 22 Jan 2026 16:20:48 +0800
Subject: [PATCH 1/1] gpio: Add support for retrieving gpio number via line
 name

Currently, the application requires GPIOs to be configured using
hardcoded integer IDs or alphanumeric pins. This commit introduces the
ability to configure GPIOs using their line names (as defined in the
device tree), which improves portability across different kernel
versions and hardware.

The implementation consists of:
1. A new `getGlobalGpioNumberFromName` function that uses `libgpiod`
   to find the line's chip and offset.
2. Logic to iterate through sysfs to find the matching gpiochip and
   retrieve its base number to calculate the gpio number .
3. Updates to `main.cpp` to parse the "line_name" key from the JSON
   configuration.

Change-Id: Ic12ab46e5b25f2f409aa501f1bd518cf2aecdc24
Signed-off-by: Mike Hsieh <mike_hsieh@quantatw.com>
---
 inc/gpio.hpp              |  2 ++
 meson.build               |  2 ++
 src/gpio.cpp              | 41 +++++++++++++++++++++++++++++++++++++++
 src/main.cpp              | 12 ++++++++++++
 subprojects/libgpiod.wrap | 13 +++++++++++++
 5 files changed, 70 insertions(+)
 create mode 100644 subprojects/libgpiod.wrap

diff --git a/inc/gpio.hpp b/inc/gpio.hpp
index 2351da9..3f6aaaa 100644
--- a/inc/gpio.hpp
+++ b/inc/gpio.hpp
@@ -70,6 +70,8 @@ int configGroupGpio(ButtonConfig& buttonCfg);
 int configGpio(GpioInfo& gpioConfig, ButtonConfig& buttonIFConfig);
 
 uint32_t getGpioNum(const std::string& gpioPin);
+
+uint32_t getGlobalGpioNumberFromName(const std::string& lineName);
 // Set gpio state based on polarity
 void setGpioState(int fd, GpioPolarity polarity, GpioState state);
 // Get gpio state based on polarity
diff --git a/meson.build b/meson.build
index 10f83fa..11966c8 100644
--- a/meson.build
+++ b/meson.build
@@ -32,6 +32,7 @@ phosphor_dbus_interfaces_dep = dependency('phosphor-dbus-interfaces')
 phosphor_logging_dep = dependency('phosphor-logging')
 sdbusplus_dep = dependency('sdbusplus')
 sdeventplus_dep = dependency('sdeventplus')
+gpiodcxx_dep = dependency('libgpiodcxx', default_options: ['bindings=cxx'])
 
 deps = [
     sdbusplus_dep,
@@ -40,6 +41,7 @@ deps = [
     nlohmann_json_dep,
     gpioplus_dep,
     sdeventplus_dep,
+    gpiodcxx_dep,
 ]
 
 sources_buttons = [
diff --git a/src/gpio.cpp b/src/gpio.cpp
index d58207e..ea898d8 100644
--- a/src/gpio.cpp
+++ b/src/gpio.cpp
@@ -21,6 +21,7 @@
 #include <fcntl.h>
 #include <unistd.h>
 
+#include <gpiod.hpp>
 #include <gpioplus/utility/aspeed.hpp>
 #include <nlohmann/json.hpp>
 #include <phosphor-logging/lg2.hpp>
@@ -126,6 +127,46 @@ uint32_t getGpioNum(const std::string& gpioPin)
     return getGpioBase() + offset;
 }
 
+uint32_t getGlobalGpioNumberFromName(const std::string& lineName)
+{
+    // Revise with c++ api
+    gpiod::line line = gpiod::find_line(lineName);
+
+    if (!line)
+    {
+        lg2::error("Failed to find the {GPIO}", "GPIO", lineName);
+        throw std::runtime_error("Could not find GPIO line!");
+    }
+
+    uint32_t offset = line.offset();
+    std::string chip_name = line.get_chip().name();
+    uint32_t base = 0;
+    bool found = false;
+
+    for (const auto& entry : fs::directory_iterator(gpioDev))
+    {
+        // Check if this sysfs entry corresponds to our chip
+        if (fs::exists(entry.path() / "device" / chip_name))
+        {
+            std::ifstream baseStream(entry.path() / "base");
+            if (baseStream >> base)
+            {
+                found = true;
+                break;
+            }
+        }
+    }
+
+    if (!found)
+    {
+        lg2::error("Could not find GPIO base for chip {CHIP}", "CHIP",
+                   chip_name);
+        throw std::runtime_error("Could not find GPIO base!");
+    }
+
+    return base + offset;
+}
+
 int configGroupGpio(ButtonConfig& buttonIFConfig)
 {
     int result = 0;
diff --git a/src/main.cpp b/src/main.cpp
index c715d5b..1990fc6 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -107,6 +107,12 @@ int main(void)
                     // When "pin" key is used, parse as alphanumeric
                     gpioCfg.number = getGpioNum(config.at("pin"));
                 }
+                else if (config.contains("line_name"))
+                {
+                    // When "line_name" key is used, parse as string
+                    gpioCfg.number =
+                        getGlobalGpioNumberFromName(config.at("line_name"));
+                }
                 else
                 {
                     // Without "pin", "num" is assumed and parsed as an integer
@@ -128,6 +134,12 @@ int main(void)
                 // When "pin" key is used, parse as alphanumeric
                 gpioCfg.number = getGpioNum(gpioConfig.at("pin"));
             }
+            else if (gpioConfig.contains("line_name"))
+            {
+                // When "line_name" key is used, parse as string
+                gpioCfg.number =
+                    getGlobalGpioNumberFromName(gpioConfig.at("line_name"));
+            }
             else
             {
                 // Without "pin", "num" is assumed and parsed as an integer
diff --git a/subprojects/libgpiod.wrap b/subprojects/libgpiod.wrap
new file mode 100644
index 0000000..f650e74
--- /dev/null
+++ b/subprojects/libgpiod.wrap
@@ -0,0 +1,13 @@
+[wrap-file]
+directory = libgpiod-1.6.3
+source_url = https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/snapshot/libgpiod-1.6.3.tar.gz
+source_filename = libgpiod-1.6.3.tar.gz
+source_hash = eb446070be1444fd7d32d32bbca53c2f3bbb0a21193db86198cf6050b7a28441
+patch_filename = libgpiod_1.6.3-1_patch.zip
+patch_url = https://wrapdb.mesonbuild.com/v2/libgpiod_1.6.3-1/get_patch
+patch_hash = 76821c637073679a88f77593c6f7ce65b4b5abf8c998f823fffa13918c8761df
+
+[provide]
+libgpiod = gpiod_dep
+libgpiodcxx = gpiodcxx_dep
+
-- 
2.31.1

